wgengine/magicsock/relaymanager.go

```diff
@@ -4,48 +4,283 @@
+	challenge [32]byte
+	disco     key.DiscoPublic
+	from      netip.AddrPort
+	vni       uint32
+	at        time.Time
+}
+
+// relayEndpointAllocWork serves to track in-progress relay endpoint allocation
+// for an [*endpoint]. This structure is immutable once initialized.
+type relayEndpointAllocWork struct {
+	// ep is the [*endpoint] associated with the work
+	ep *endpoint
+	// cancel() will signal all associated goroutines to return
+	cancel context.CancelFunc
+	// wg.Wait() will return once all associated goroutines have returned
+	wg *sync.WaitGroup
+}
+
+// init initializes [relayManager] if it is not already initialized.
+func (r *relayManager) init() {
+	r.initOnce.Do(func() {
+		r.discoInfoByServerDisco = make(map[key.DiscoPublic]*discoInfo)
+		r.allocWorkByEndpoint = make(map[*endpoint]*relayEndpointAllocWork)
+		r.allocateHandshakeCh = make(chan *endpoint)
+		r.allocateWorkDoneCh = make(chan relayEndpointAllocWorkDoneEvent)
+		r.cancelWorkCh = make(chan *endpoint)
+		r.newServerEndpointCh = make(chan newRelayServerEndpointEvent)
+		r.rxChallengeCh = make(chan relayHandshakeChallengeEvent)
+		r.rxCallMeMaybeViaCh = make(chan *disco.CallMeMaybeVia)
+		r.runLoopStoppedCh = make(chan struct{}, 1)
+		go r.runLoop()
+	})
 }
 
 // discoInfo returns a [*discoInfo] for 'serverDisco' if there is an
 // active/ongoing handshake with it, otherwise it returns nil, false.
-func (h *relayManager) discoInfo(serverDisco key.DiscoPublic) (_ *discoInfo, ok bool) {
-	h.mu.Lock()
-	defer h.mu.Unlock()
-	h.initLocked()
-	di, ok := h.discoInfoByServerDisco[serverDisco]
+func (r *relayManager) discoInfo(serverDisco key.DiscoPublic) (_ *discoInfo, ok bool) {
+	r.discoInfoMu.Lock()
+	defer r.discoInfoMu.Unlock()
+	di, ok := r.discoInfoByServerDisco[serverDisco]
 	return di, ok
 }
 
-func (h *relayManager) handleCallMeMaybeVia(dm *disco.CallMeMaybeVia) {
-	h.mu.Lock()
-	defer h.mu.Unlock()
-	h.initLocked()
-	// TODO(jwhited): implement
+func (r *relayManager) handleCallMeMaybeVia(dm *disco.CallMeMaybeVia) {
+	relayManagerInputEvent(r, nil, &r.rxCallMeMaybeViaCh, dm)
+}
+
+func (r *relayManager) handleBindUDPRelayEndpointChallenge(dm *disco.BindUDPRelayEndpointChallenge, di *discoInfo, src netip.AddrPort, vni uint32) {
+	relayManagerInputEvent(r, nil, &r.rxChallengeCh, relayHandshakeChallengeEvent{challenge: dm.Challenge, disco: di.discoKey, from: src, vni: vni, at: time.Now()})
+}
+
+// relayManagerInputEvent initializes [relayManager] if necessary, starts
+// relayManager.runLoop() if it is not running, and writes 'event' on 'eventCh'.
+//
+// [relayManager] initialization will make `*eventCh`, so it must be passed as
+// a pointer to a channel.
+//
+// 'ctx' can be used for returning when runLoop is waiting for the caller to
+// return, i.e. the calling goroutine was birthed by runLoop and is cancelable
+// via 'ctx'. 'ctx' may be nil.
+func relayManagerInputEvent[T any](r *relayManager, ctx context.Context, eventCh *chan T, event T) {
```

Comment:
this ends up being monomorphized per event type which feels a bit unfortunate.

it feels slightly complicated to fix, i feel let down by go channels and types again here, but, if every channel was buffered (at least 1 buffered), then `activeWork` could also sum `len()` for each chan as well as part of its computation, and the caller could push on to the channel then call a "maybeStart". I think I would drop the ctx input here as well, to make that easier.

I also might consider just having this goroutine live "forever" or a lifetime tied to magicsock.Conn or similar, as it also simplifies the reactor here and isn't too expensive if there's only one (it may cost less memory than the monomorphization of this code if init() gets inlined).

Reply:
> this ends up being monomorphized per event type which feels a bit unfortunate.

There are 6 event types currently and that is likely a complete set for what `relayManager` needs to do.

> if every channel was buffered (at least 1 buffered), then activeWork could also sum len() for each chan as well as part of its computation, and the caller could push on to the channel then call a "maybeStart". I think I would drop the ctx input here as well, to make that easier.

What would maybeStart implementation look like? I'm guessing the synchronization might be more difficult to reason about vs as written.

> it may cost less memory than the monomorphization of this code if init() gets inlined

Alternatively init() could get pulled out and duplicated at the call sites.

Reply:
yeah, i had a play with it. the language is fundamentally too weak to implement this efficiently, you end up needing tagged unions to really clean things up, and i don't want to suggest going down the reflect route. the one slightly simplified case I had introduces risks of deadlocks if it's not held perfectly, and that seems strictly worse.